[Quant] [Inductor] Enable the Inductor Lowering of QConv2d post op SiLU

zhuhaozhe opened the pull request:
**Summary**
Enable the fusion pattern of `QConv2d -> silu` lowering to `swish` as `QConv2d` post operator. 

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_silu_cpu
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_silu_int8_mixed_bf16_cpu
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_silu
```


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122374
* #122373
* __->__ #122268
* #122267
* #122266



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

zhuhaozhe committed:
[Quant] [Inductor] Enable the Inductor Lowering of QConv2d post op SiLU

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -656,6 +658,15 @@ def __init__(self, op_name: str, scalars_attr=None, algorithm_attr=None):
+                    2,
+                    is_bf16,
+                ),
+                dtype=original_pattern_output_dtype,
```
For `int-mixed-bf16` and `int8 output`, here we will have the `to_fp32` between `qconv->mul/sigmod` (from `silu`). I think the pattern should have `to_fp32` between `silu` and output's `quant`. Should we keep using `torch.float32` instead of `original_pattern_output_dtype` here?

leslie-fang-intel replied:
We may also check `qconv2d_unary_matcher_nodes` in UT to ensure the correct number of nodes in pattern has been matched.

zhuhaozhe replied:
Thanks for reminder. Updated under your advise.

zhuhaozhe committed:
Update on "[Quant] [Inductor] Enable the Inductor Lowering of QConv2d post op SiLU"

**Summary**
Enable the fusion pattern of `QConv2d -> silu` lowering to `swish` as `QConv2d` post operator. 

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_silu_cpu
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_silu_int8_mixed_bf16_cpu
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_silu
```





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -558,6 +558,7 @@ def _qconv2d_unary_cpu_test_helper(
         self,
         int8_mixed_bf16=False,
         unary_op=torch.nn.ReLU(),
+        qconv2d_unary_matcher_nodes=None,
```
May pass in `matcher_check_fn`?

jgong5 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -656,6 +658,15 @@ def __init__(self, op_name: str, scalars_attr=None, algorithm_attr=None):
+                    _silu_fusion,
+                    get_dequantize_qconv_pt2e_pattern(1 if is_bf16 else 2),
+                    2,
+                    is_bf16,
```
How about other fusions above?

zhuhaozhe replied:
Updated for `hardsiwsh` and `hardtanh` https://github.com/pytorch/pytorch/pull/122374

jgong5 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -680,6 +691,12 @@ def __init__(self, op_name: str, scalars_attr=None, algorithm_attr=None):
+                _silu_fusion,
+                get_dequantize_qconv_pt2e_pattern(1 if is_bf16 else 2),
+                2,
+                is_bf16,
```
ditto

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe commented:
@pytorchbot merge
